pkg/pillar/pubsub/socketdriver/publish.go

```diff
@@ -116,9 +133,49 @@ func (s *Publisher) Load() (map[string][]byte, int, error) {
 		}
 		items[key] = sb
 	}
+
+	// Try to recover lost items (have backup but missing the original).
+	if s.persistent {
+		const backupSuffix = ".json.bak"
+		var keysToRecover []string
+		for _, file := range files {
+			if strings.HasSuffix(file.Name(), backupSuffix) {
+				key := strings.TrimSuffix(file.Name(), backupSuffix)
+				if _, loaded := items[key]; !loaded {
+					keysToRecover = append(keysToRecover, key)
+				}
+			}
+		}
+		for _, key := range keysToRecover {
+			sb, err := s.RecoverFromBackup(key)
+			if err != nil {
+				s.log.Errorf("Failed to recover %s/%s: %v", s.name, key, err)
+				continue
+			}
+			items[key] = sb
+			s.log.Warnf("Using backup of %s/%s", s.name, key)
+		}
+	}
 	return items, restartCounter, err
 }
 
+// RecoverFromBackup tries to recover possibly corrupted item from a backup file.
+// Typically called when loaded value cannot be unmarshalled into a structure,
+// thus indicating file data corruption.
+func (s *Publisher) RecoverFromBackup(key string) ([]byte, error) {
+	origFileName := s.dirName + "/" + key + ".json"
+	backupFileName := origFileName + ".bak"
+	err := fileutils.CopyFile(backupFileName, origFileName)
```

Comment:
If we get a second power outage when we do this Copy we can still end up with corruption (but we'd still have the backup).

I wonder we it would be slightly better to do a ReadFuke of the backupFilename, and then do a WriteRename of that content to origFileName. But ideally we'd want to do that after the ParseTemplate which means that the WriteRename should be in the caller.

Reply:
Maybe this also needs to be a Publish call so that any subscribers get notified about the changes caused by applying the content in the .bak file.

Reply:
Done: recovery procedure is now Read followed by WriteRename.

Reply:
Thanks